packages/legend-query-builder/src/components/shared/BasicValueSpecificationEditor.tsx

```diff
@@ -628,153 +623,271 @@ const stringifyValue = (values: ValueSpecification[]): string => {
+  isLoading: boolean;
+  reloadValues:
+    | DebouncedFunc<(inputValue: string) => GeneratorFn<void>>
+    | undefined;
+  cleanUpReloadValues?: () => void;
+}
 
-  if (!parseData) {
-    return;
-  }
+const PrimitiveCollectionInstanceValueEditor = observer(
+  (props: {
+    valueSpecification: CollectionInstanceValue;
+    expectedType: Type;
+    saveEdit: () => void;
+    selectorConfig?: BasicValueSpecificationEditorSelectorConfig | undefined;
+    observerContext: ObserverContext;
+  }) => {
+    const {
+      valueSpecification,
+      expectedType,
+      saveEdit,
+      selectorConfig,
+      observerContext,
+    } = props;
 
-  if (expectedType instanceof PrimitiveType) {
-    switch (expectedType.path) {
-      case PRIMITIVE_TYPE.STRING: {
-        result = uniq(parseData)
-          .map((item): PrimitiveInstanceValue | undefined => {
-            const primitiveInstanceValue = new PrimitiveInstanceValue(
-              GenericTypeExplicitReference.create(
-                new GenericType(expectedType),
-              ),
-            );
-            instanceValue_setValues(
-              primitiveInstanceValue,
-              [item.toString()],
-              obseverContext,
-            );
-            return primitiveInstanceValue;
-          })
-          .filter(isNonNullable);
-        break;
+    // local state and variables
+    const applicationStore = useApplicationStore();
+    const inputRef = useRef(null);
+    const [inputValue, setInputValue] = useState('');
+    const [inputValueIsError, setInputValueIsError] = useState(false);
+    const [selectedOptions, setSelectedOptions] = useState<
+      { label: string; value: string }[]
+    >(
+      valueSpecification.values
+        .map((valueSpec) => getValueSpecificationStringValue(valueSpec))
+        .filter(isNonEmptyString)
+        .map((value) => ({
+          label: value,
+          value,
+        })),
+    );
+
+    // typehead search setup
+    const isTypeaheadSearchEnabled =
+      expectedType === PrimitiveType.STRING && Boolean(selectorConfig);
+    const reloadValuesFunc = isTypeaheadSearchEnabled
+      ? selectorConfig?.reloadValues
+      : undefined;
+    const cleanUpReloadValuesFunc = isTypeaheadSearchEnabled
+      ? selectorConfig?.cleanUpReloadValues
+      : undefined;
+    const isLoading = isTypeaheadSearchEnabled
+      ? selectorConfig?.isLoading
+      : undefined;
+    const queryOptions =
+      isTypeaheadSearchEnabled && selectorConfig?.values?.length
+        ? selectorConfig.values.map((e) => ({
+            value: e,
+            label: e.toString(),
+          }))
+        : undefined;
+    const noMatchMessage =
+      isTypeaheadSearchEnabled && isLoading ? 'Loading...' : undefined;
+
+    // helper functions
+    const buildOptionForValueSpec = (
+      value: ValueSpecification,
+    ): { label: string; value: string } => ({
+      label: guaranteeNonNullable(getValueSpecificationStringValue(value)),
```

Comment:
maybe we could create a local variable with value `getValueSpecificationStringValue(value)` 
so that this calculation will only be done once

Reply:
Good point! Made the change